src/main/flight/safetydroneflip.c

```diff
@@ -0,0 +1,176 @@
+    motor[0] = Motorthrotle-500; //right back
+    motor[1] = Motorthrotle;   //right front
+    motor[2] = Motorthrotle-500;   //left back
+    motor[3] = Motorthrotle;   //left front
+    }
+}
+
+bool batteryIsCritical(void){
+    return batteryCriticalStatus;
+}
+
+uint16_t GetThrottle(int reset) //get the throttle for the drone
+  {
+    static uint16_t Motorthrottle = 950;
+    static bool ValueUp = false; //sees if the value for Motorthrottle hase gone up
+    static unsigned long Pervious_millis = 0;
+
+    if (reset == 1){ //if the drone is upside down the function is reset
+        Motorthrottle = 950;
+        Pervious_millis = millis();
+        return 0;
+    }
+
+    if ((millis()-Pervious_millis) <= 3000) //tries to flip for 3 sec after this the drone will up the motor throtle
+    {
+        if (ValueUp == false)
+        {
+            Motorthrottle += 50;
+            if (Motorthrottle >= 1500)
+            {
+                Motorthrottle = 1500;
+
+            }
+        ValueUp = true;
+        }
+        return Motorthrottle;
+    }
+    if (((millis()-Pervious_millis)  > 3000)&& ((millis()-Pervious_millis)  < 6000)) //after 3 sec the motors will turn of for 3 sec
+    {
+        ValueUp = false;
+        return 0;
+    }
+    if ((millis()-Pervious_millis) >= 6000) //after total of 6 sec time is reset
+    {
+        Pervious_millis = millis();
+        return 0;
+    }
+    return 0;
+}
+
+
+void safetydroneflipMain (void){
+
+    float orientation = isupsidedown(); //used to check in matrix if the drone is upside down <-.5 is fine
+
+    static uint16_t Motorthrottle = 950; //motor throtle given to the motors of the drone
+   
+    uint16_t Voltage = getBatteryVoltageLatest();
+    static uint16_t lowestvalue = 0;
+    static uint16_t Previouslowestvalue = 500;
+    
+    static float voltageSmoothed = 0.0;
+    const float Alpha = 0.15;
+    voltageSmoothed = (1.0-Alpha)*voltageSmoothed+Alpha*Voltage;
+   
+    static unsigned long previousMS = 0;
+    static unsigned long previousMsSafeVoltage = 0;
+
+    if ( (millis()-previousMS) <= 30000){
+       if  (voltageSmoothed < Previouslowestvalue){
+           Previouslowestvalue = voltageSmoothed;
+       }   
+    }
+    if ((millis()-previousMS) > 30000){
+        previousMS = millis();
+        lowestvalue = Previouslowestvalue;
+        Previouslowestvalue = 500;
+    }
+
+    if ((lowestvalue >= UNSAFE_VALUE) && (batteryCriticalStatus == true)){
+        previousMsSafeVoltage = millis(); 
+    }
+
+    if ((lowestvalue >= UNSAFE_VALUE) || (batteryCriticalStatus == true)|| (voltageSmoothed > UNSAFE_VALUE_HIGH))
+    { 
+        if (orientation < -0.5) //if upside down
+        {
+            if (setReverse) {        
+                Motors_normal();
+            }
+            Motors_out();
+            GetThrottle(1);
+            batteryCriticalStatus = false;
+            Motorthrottle = 950;
+        }
+        else //not upside down or sensor not working
```

Comment:
If the sensor is not working, or the IMU is wrongly calibrated, it will not flip right?